planner/core/rule_join_reorder.go

```diff
@@ -107,6 +116,47 @@ func (s *joinReOrderSolver) optimizeRecursive(ctx sessionctx.Context, p LogicalP
 			joinTypes:  joinTypes,
 		}
 
+		joinGroupNum := len(curJoinGroup)
+		var leadingJoinGroup []LogicalPlan
+		if hintInfo != nil {
+			for _, hintTbl := range hintInfo.leadingJoinOrder {
+				for i, joinGroup := range curJoinGroup {
+					tableAlias := extractTableAlias(joinGroup, joinGroup.SelectBlockOffset())
+					if tableAlias == nil {
+						continue
+					}
+					if hintTbl.dbName.L == tableAlias.dbName.L && hintTbl.tblName.L == tableAlias.tblName.L && hintTbl.selectOffset == tableAlias.selectOffset {
+						leadingJoinGroup = append(leadingJoinGroup, joinGroup)
+						curJoinGroup = append(curJoinGroup[:i], curJoinGroup[i+1:]...)
+						break
+					}
+				}
+			}
+			var errMsg string
+			if len(leadingJoinGroup) != len(hintInfo.leadingJoinOrder) {
+				errMsg = "leading hint is inapplicable, check if the leading hint table is valid"
+			} else if joinGroupNum <= ctx.GetSessionVars().TiDBOptJoinReorderThreshold {
+				errMsg = "leading hint is inapplicable for the DP join reorder algorithm"
+			}
+			if len(errMsg) > 0 {
+				curJoinGroup = append(curJoinGroup, leadingJoinGroup...)
+				leadingJoinGroup = nil
```

Comment:
Seems that the leading hint cannot be working if the join group number is less than `TiDBOptJoinReorderThreshold`?

Reply:
fixed. PTAL